Userland/Libraries/LibCompress/Lzma.cpp

```diff
@@ -0,0 +1,604 @@
+                }
+
+                u8 byte;
+                auto read_bytes = TRY(m_output_buffer.read_with_seekback({ &byte, sizeof(byte) }, m_rep0 + 1));
+                VERIFY(read_bytes.size() == sizeof(byte));
+
+                auto written_bytes = m_output_buffer.write({ &byte, sizeof(byte) });
+                VERIFY(written_bytes == sizeof(byte));
+                m_total_decoded_bytes++;
+
+                real_length--;
+            }
+
+            return {};
+        };
+
+        // If we have a leftover part of a repeating match, we should finish that first.
+        if (m_leftover_match_length.has_value()) {
+            TRY(copy_match_to_buffer(m_leftover_match_length.release_value()));
+            continue;
+        }
+
+        // "The decoder uses the following code flow scheme to select exact
+        //  type of LITERAL or MATCH:
+        //
+        //  IsMatch[state2] decode
+        //   0 - the Literal"
+        if (TRY(decode_bit_with_probability(m_is_match_probabilities[state2])) == 0) {
+            // "At first the LZMA decoder must check that it doesn't exceed
+            //  specified uncompressed size."
+            if (m_options.uncompressed_size.has_value() && m_options.uncompressed_size.value() <= m_total_decoded_bytes)
+                return Error::from_string_literal("Exceeded uncompressed archive size while decoding literal symbol");
+
+            // "Then it decodes literal value and puts it to sliding window."
+            TRY(decode_literal_to_output_buffer());
+
+            // "Then the decoder must update the "state" value."
+            update_state_after_literal();
+            continue;
+        }
+
+        // " 1 - the Match
+        //     IsRep[state] decode
+        //       0 - Simple Match"
+        if (TRY(decode_bit_with_probability(m_is_rep_probabilities[m_state])) == 0) {
+            // "The distance history table is updated with the following scheme:"
+            m_rep3 = m_rep2;
+            m_rep2 = m_rep1;
+            m_rep1 = m_rep0;
+
+            // "The zero-based length is decoded with "LenDecoder"."
+            u16 normalized_length = TRY(decode_normalized_match_length(m_length_decoder));
+
+            // "The state is update with UpdateState_Match function."
+            update_state_after_match();
+
+            // "and the new "rep0" value is decoded with DecodeDistance."
+            m_rep0 = TRY(decode_normalized_match_distance(normalized_length));
+
+            // "If the value of "rep0" is equal to 0xFFFFFFFF, it means that we have
+            //  "End of stream" marker, so we can stop decoding and check finishing
+            //  condition in Range Decoder"
+            if (m_rep0 == 0xFFFFFFFF) {
+                if (m_range_decoder_code != 0)
+                    return Error::from_string_literal("Found 'end of stream' marker, but range decoder is in unfinished state");
+
+                m_found_end_of_stream_marker = true;
+                continue;
```

Comment:
We set `m_found_end_of_stream_marker` (this is the only place where we do so), and immediately continue the loop, at which point we break out of it using the loop end condition at the top. We then go into the post-loop check for the code validity if we reached the end of stream marker.
AKA: This check for the final code value is duplicated, we should remove one of the copies.